File: 1197-parsing-a-boolean-expression/1197-parsing-a-boolean-expression.cpp
class Solution {
    void evaluate(std::stack<char>& optr, std::stack<char>& operand) {
        char op = optr.top();  // Get the character at the top
        optr.pop();  // Remove the operator from the stack
        bool ans;

        if (op == '|') {
            ans = false;  // For OR, initial value should be false
            while (operand.top() != '(') {
                ans |= (operand.top() == 't');
                operand.pop();
            }
            operand.pop();  // Pop the '('
        } else if (op == '&') {
            ans = true;  // For AND, initial value should be true
            while (operand.top() != '(') {
                ans &= (operand.top() == 't');
                operand.pop();
            }
            operand.pop();  // Pop the '('
        } else if (op == '!') {
            ans = (operand.top() == 't');
            ans = !ans;  // Negate the value
            operand.pop();  // Pop the value
            operand.pop();  // Pop the '('
        }

        if (ans) operand.push('t');  // Push true result back to operand stack
        else operand.push('f');  // Push false result back to operand stack
    }

public:
    bool parseBoolExpr(std::string expression) {
        std::stack<char> optr, operand;
        int idx = 0;

        while (expression[idx]) {
            if (expression[idx] == '!' || expression[idx] == '&' || expression[idx] == '|') {
                optr.push(expression[idx]);
            } else if (expression[idx] == 't' || expression[idx] == 'f') {
                operand.push(expression[idx]);  // Push operand
            } else if (expression[idx] == '(') {
                operand.push(expression[idx]);  // Pushing opening parenthesis
            } else if (expression[idx] == ')') {
                evaluate(optr, operand);  // Evaluate when closing parenthesis is encountered
            }
            idx++;
        }

        return operand.top() == 't';  // Final result check
    }
};

File: 1197-parsing-a-boolean-expression/1197-parsing-a-boolean-expression.rs
use std::collections::VecDeque;

impl Solution {
    pub fn parse_bool_expr(expression: String) -> bool {
        let mut stk = VecDeque::new();  // Stack to hold characters and operators

        // Iterate over each character in the expression
        for c in expression.chars() {
            if c != ')' && c != ',' {
                stk.push_back(c);  // Push valid characters to the stack
            } else if c == ')' {  // When ')' is encountered, evaluate subexpression
                let mut exp = Vec::new();  // Vector to hold boolean values
                
                // Pop characters until '(' is found, collect 't' or 'f' values
                while let Some(&top) = stk.back() {
                    if top == '(' {
                        break;
                    }
                    stk.pop_back();
                    exp.push(top == 't');
                }
                
                stk.pop_back();  // Pop the '(' from the stack
                
                if let Some(&op) = stk.back() {
                    stk.pop_back();
                    let mut v = exp[0];  // Initialize result with the first value
                    
                    // Perform the corresponding logical operation
                    if op == '&' {  // AND operation
                        for &b in &exp {
                            v &= b;
                        }
                    } else if op == '|' {  // OR operation
                        for &b in &exp {
                            v |= b;
                        }
                    } else {  // NOT operation
                        v = !v;
                    }
                    
                    // Push the result back to the stack as 't' or 'f'
                    if v {
                        stk.push_back('t');
                    } else {
                        stk.push_back('f');
                    }
                }
            }
        }
        
        // Return the final result from the stack
        stk.back() == Some(&'t')
    }
}
